test/PushNotificationTests/BaseTestSuite.cpp

```diff
@@ -0,0 +1,264 @@
+
+void BaseTestSuite::ClassCleanup()
+{
+    ::Test::Bootstrap::Cleanup();
+}
+
+void BaseTestSuite::MethodCleanup()
+{
+    if (m_registered)
+    {
+        PushNotificationManager::Default().UnregisterAll();
+        m_registered = false;
+    }
+}
+
+HRESULT BaseTestSuite::ChannelRequestHelper(IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& channelOperation)
+{
+    if (channelOperation.wait_for(c_timeout) != winrt::Windows::Foundation::AsyncStatus::Completed)
+    {
+        channelOperation.Cancel();
+        return HRESULT_FROM_WIN32(ERROR_TIMEOUT); // timed out or failed
+    }
+
+    auto result{ channelOperation.GetResults() };
+    auto status{ result.Status() };
+    if (status != PushNotificationChannelStatus::CompletedSuccess)
+    {
+        return result.ExtendedError(); // did not produce a channel
+    }
+
+    result.Channel().Close();
+    return S_OK;
+}
+
+void BaseTestSuite::ChannelRequestUsingNullRemoteId()
+{
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        VERIFY_THROWS_HR(PushNotificationManager::Default().CreateChannelAsync(winrt::guid()).get(), E_INVALIDARG);
+    }
+    else
+    {
+        auto channelOperation{ PushNotificationManager::Default().CreateChannelAsync(winrt::guid()) };
+        VERIFY_ARE_EQUAL(ChannelRequestHelper(channelOperation), E_FAIL);
+    }
+}
+
+void BaseTestSuite::ChannelRequestUsingRemoteId()
+{
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        auto channelOperation{ PushNotificationManager::Default().CreateChannelAsync(c_azureRemoteId) };
+        VERIFY_SUCCEEDED(ChannelRequestHelper(channelOperation));
+    }
+    else
+    {
+        auto channelOperation{ PushNotificationManager::Default().CreateChannelAsync(c_azureRemoteId) };
+        VERIFY_ARE_EQUAL(ChannelRequestHelper(channelOperation), E_FAIL);
+    }
+}
+
+// Currently failing - https://github.com/microsoft/WindowsAppSDK/issues/2392
+void BaseTestSuite::MultipleChannelClose()
+{
+    auto channelOperation{ PushNotificationManager::Default().CreateChannelAsync(c_azureRemoteId) };
+    if (channelOperation.wait_for(c_timeout) != winrt::Windows::Foundation::AsyncStatus::Completed)
+    {
+        channelOperation.Cancel();
+        VERIFY_FAIL(L"Channel request hit timeout.");
+    }
+
+    auto result{ channelOperation.GetResults() };
+    VERIFY_ARE_EQUAL(result.Status(), PushNotificationChannelStatus::CompletedSuccess);
+
+    result.Channel().Close();
+    VERIFY_THROWS_HR(result.Channel().Close(), WPN_E_CHANNEL_CLOSED);
+}
+
+void BaseTestSuite::VerifyRegisterAndUnregister()
+{
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    PushNotificationManager::Default().Unregister();
+}
+
+void BaseTestSuite::VerifyUnregisterFails()
+{
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        VERIFY_THROWS_HR(PushNotificationManager::Default().Unregister(), E_UNEXPECTED);
```

Comment:
Ooooooh. That's neat. I didn't know TAEF had VERIFY_THROWS*()

Reply:
It doesn't exist... but I added a define in pch.h that has the same functionality as VERIFY_THROWS_HR.

#define VERIFY_THROWS_HR(expression, hr)        \
            VERIFY_THROWS_SPECIFIC(expression,          \
                winrt::hresult_error,                   \
                [&](winrt::hresult_error e) -> bool     \
                {                                       \
                    return (e.code() == hr);    \
                })